# 29회 ADP 복원 

# 기계학습 (60점)

## 데이터 설명
- 데이터 출처 : https://www.data.go.kr/data/15094266/fileData.do 후처리
- 데이터 링크 : https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/29/p1.csv
- 데이터 설명 : 대구도시공사_빅데이터_영구임대아파트 입주자 퇴거여부 데이터, 고유번호를 가진 계약자와 특정 아파트에 대해 매년 퇴거여부를 기록한 데이터

### 1-1  계약자고유번호를 기준으로 거주연도 별 여러개의 데이터가 쌓여 있다. 각 계약자고유번호에 대해 가장 최신의 거주연도 행만 남겨라.

In [7]:
import ssl
import pandas as pd
import urllib
context = ssl._create_unverified_context()

url1 = 'https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/29/p1.csv'
encoded_url1 = urllib.parse.quote(url1, safe=':/')

response = urllib.request.urlopen(encoded_url1, context=context)
df1 = pd.read_csv(response, encoding='euc-kr')
df1

순번 계약구분  재계약횟수  거주개월    아파트 이름  아파트 ID  아파트 평점  호실고유번호   층  평형대  \
0          1   유효     10   222     강남아파트       5     7.0   14520   1   12   
1          1   유효     10   222     강남아파트       5     7.0   14520   1   12   
2          1   유효     10   222     강남아파트       5     7.0   14520   1   12   
3          1   유효     10   222     강남아파트       5     7.0   14520   1   12   
4          1   유효     10   222     강남아파트       5     7.0   14520   1   12   
...      ...  ...    ...   ...       ...     ...     ...     ...  ..  ...   
86899  12882   해지      8   180  지산5단지아파트       3     8.0   86891  14   15   
86900  12882   해지      8   180  지산5단지아파트       3     8.0   86891  14   15   
86901  12883   유효      2    36  지산5단지아파트       3     8.0   86891  14   15   
86902  12883   유효      2    36  지산5단지아파트       3     8.0   86891  14   15   
86903  12883   유효      2    36  지산5단지아파트       3     8.0   86891  14   15   

       ...    퇴거연도  거주연도  월세(원)   보증금(원)  대표나이  나이  성별  결혼여부  거주자 수 퇴거여부  
0      ...     NaN  2008  47100  3646000    46  33   남    미혼      3  미퇴거  
1      ...     NaN  2009  56500  4375000    46  34   남    미혼      3  미퇴거  
2      ...     NaN  2010  56500  4375000    46  35   남    미혼      3  미퇴거  
3      ...     NaN  2011  69900  5408000    46  36   남    미혼      3  미퇴거  
4      ...     NaN  2012  69900  5408000    46  37   남    미혼      3  미퇴거  
...    ...     ...   ...    ...      ...   ...  ..  ..   ...    ...  ...  
86899  ...  2018.0  2017  49500  3055000    60  56   남    기혼      2  미퇴거  
86900  ...  2018.0  2018  49500  3055000    60  57   남    기혼      2   퇴거  
86901  ...     NaN  2018  45000  2365000    52  49   남    미혼      1  미퇴거  
86902  ...     NaN  2019  45000  2365000    52  50   남    미혼      1  미퇴거  
86903  ...     NaN  2020  45000  2365000    52  51   남    미혼      1  미퇴거  

[86904 rows x 23 columns]

### 1-2 결측치를 처리하라

### 1-3 이상치를 처리하라

### 2-1 재계약 횟수의 중앙값을 기준으로 중앙값보다 크거나 같으면 '높음', 작으면 '낮음' 으로 재계약 횟수 이분 변수를 구성하시오.

### 2-2 차원축소의 필요성을 논하고, 필요에 따라 차원을 축소하고 불필요하다면 그 근거를 논하시오.

### 3-1 2번에서 진행한 재계약 횟수 이분변수를 기준으로 세그먼트를 구분하고 각 세그먼트의 특징을 분석하시오.

### 3-2. 재계약횟수 변수를 종속변수로 하는 회귀 분석 모델을 두 가지 이상의 방법론을 통해 수행하고 최종 모델을 결정하시오. 재계약횟수 이분변수를 종속변수로 하는 분류 분석을 두가지 이상의 방법론을 통해 수행하고 최종 모델을 결정한 후 그 이유를 작성하시오 

### 3-3 최종 채택한 모델에서 각각 변수 중요도를 확인하고 각 변수의 차이점을 설명하시오

### 3-4 해당 데이터 분석결과로 얻을 수 있는 점 제시

## 데이터 설명
- 데이터 출처 : https://www.kaggle.com/datasets/pschale/mlb-pitch-data-20152018 데이터를 후처리
- 데이터 링크 : https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/29/p2_.csv
- 데이터 설명 : 

A 야구구단의 시합 결과중 일부를 나타낸다.    
각 행은 하나의 경기를 의미(game_id)하며 9회차(= 9이닝) 동안 1번타자, 2번타자의 출루 정보를 의미한다.     
각 컬럼에 대한 설명은 아래 이미지와 같으며, value값의 index는 이미지의 '수치 의미' 열을 통해 확인 할 수 있다.


<img src="https://github.com/Datamanim/datarepo/blob/main/adp/29/p1_img.png?raw=true" width="500" height="600">

In [13]:
import ssl
import pandas as pd
import urllib
context = ssl._create_unverified_context()

url2 = 'https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/29/p2_.csv'
encoded_url2 = urllib.parse.quote(url2, safe=':/')

response = urllib.request.urlopen(encoded_url2, context=context)
df2 = pd.read_csv(response, encoding='utf-8')
df2

game_id a1_1  a1_2 a2_1  a2_2  a3_1  a3_2 a4_1  a4_2  a5_1  ...  a9_2  \
0    201900016    5     5    5     5     5     5    5     5     5  ...     5   
1    201900023    6     4    5     5     2     5    1     1     6  ...     7   
2    201900103    5     6    5     5     1     9    5     4     6  ...     6   
3    201900112    5     7    6     1     5     5    1     5     1  ...     5   
4    201900131    5     1    2     5     2     5    7     2     6  ...     6   
..         ...  ...   ...  ...   ...   ...   ...  ...   ...   ...  ...   ...   
195  201902346    7     5    7     4     1     9    5     7     7  ...     1   
196  201902365    2     5    1     1     7     5    5     7     5  ...     9   
197  201902373          7    5     6     5     9    7     5     1  ...     5   
198  201902392    6     5    5     1     1     5    6     5     5  ...     7   
199  201902394    2     4    7     4     5     5    7     5     6  ...     5   

     b1  b2 b3  b4  b5  b6  b7  b8  b9  
0     0   0  0   0   0   0   0   4   2  
1     3   0  0   1   0   1   3   1   0  
2     0   0  1   1   4   1   2   0   1  
3     0   3  0   0   0   0   1   0   0  
4     0   2  4   1   2   2   0   1   0  
..   ..  .. ..  ..  ..  ..  ..  ..  ..  
195   0   1  2   0   0   0   0   0   1  
196   0   0  0   0   1   0   0   0   2  
197   0   0  0   0   0   1   0   0   0  
198   0   0  2   1   0   0   0   2   1  
199   2   1  1   0   0   4   0   0   1  

[200 rows x 28 columns]

### 4-1 각 회차별로 1번 타자의 출루 (1,2,3루타와 사사구(볼넷, 몸에맞는공))가 있는 경우에 대해 득점이 발생 했는지 확인하고자 한다. 이를 위한 전처리를 수행하라. (단, 첫 번째 혹은 두 번째 타자가 홈런을 친 경우 해당 회차 데이터는 제외한다.)     

조건1 : 득점여부를 범주형 종속변수로 한다. (1점이상 득점 :1, 무득점 :0)     
조건2 : 각 회차 2번 타자의 데이터는 원핫 인코딩한다.    
조건3 : 학습에 적절하지 않은 데이터는 제외한다.    

### 4-2 4-1 데이터에 대해 Logistic Regression을 적용하고 2번타자의 희생번트 여부에 대한 회귀 계수 검정을 하라

### 4-3 SMOTE (random_state =0 지정)를 적용하여 data imbalance를 해결하라

### 4-4 4-3에서 변환한 데이터를 기준으로 Logistic Regression을 적용하고 결과를 분석하라

### 4-5 4-3에서 변환한 데이터를 기준으로 XGBoost 적용하고 결과를 분석하라

# 통계 (40점)

### 5. TV 패널을 생산하는 공장에서 제품 A의 불량률은 0.03이다. 25개의 제품을 뽑았을 때 3개가 불량일 확률을 구하시오. (소수점 다섯 째 자리에서 반올림)

### 6. A, B 두 공장에서 생산되는 제품 중에서 양품일 확률은 15,000개 중에서 11,800개가 양품, 21,000개 중에서 16,400개가 양품이다. 두 공장의 양품 확률이 같은지를 검정하라

#### 6-1 연구가설과 귀무가설을 세워라

#### 6-2 검정 통계량을 구하고 10% 유의수준에서 검정하여라

## 7. 아래 데이터는 a,b,c,d 네 차종 각각 5회 실험 시 범퍼 파손 정도 이다. (단, 각 모집단은 정규분포를 따르며 모집단 간 등분산성을 가정한다.)

dataurl : https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/29/p7.csv

In [12]:
import ssl
import pandas as pd
import urllib
context = ssl._create_unverified_context()

url3 = 'https://raw.githubusercontent.com/Datamanim/datarepo/main/adp/29/p7.csv'
encoded_url3 = urllib.parse.quote(url3, safe=':/')

response3 = urllib.request.urlopen(encoded_url3, context=context)
df3 = pd.read_csv(response3, encoding='utf-8')
df3

name      ratio
0     A  10.993428
1     A   9.723471
2     A  11.295377
3     A  13.046060
4     A   9.531693
5     B  14.297589
6     B  19.737638
7     B  17.302304
8     B  13.591577
9     B  16.627680
10    C   7.304873
11    C   7.301405
12    C   8.362943
13    C   5.130080
14    C   5.412623
15    D  10.594281
16    D   9.467922
17    D  12.785618
18    D   9.729940
19    D   8.469241

### 7-1. 각 차종 별 범퍼 파손의 정도에 차이가 유의한지 검정하라.

#### 7-1 연구가설과 귀무가설을 세워라

#### 7-2 검정 통계량을 구하여라

#### 7-3 귀무가설을 기각하는가? 그리고 그 이유를 설명하여라. 만약 귀무가설을 기각하지 못한다면 다중비교(사후비교)를 시행하여라

### 8. L1,L2,L3 세 개의 생산라인에서 각각 13%, 37%, 50%를 생산하며 각각 2.1% , 1.1%, 3.3% 불량률을 갖는다. 불량 제품이 나왔을 때 L1 라인에서 생산되었을 확률을 구하시오. (소수점 둘째자리에서 반올림)